### PR TITLE
Bye bye spree - final PR :tada:

### PR DESCRIPTION
**Pull request by luisramos0**
#### What? Why?

This depends on (both already merged and integrated in this PR :+1: )
https://github.com/openfoodfoundation/better_spree_paypal_express/pull/9
and
https://github.com/openfoodfoundation/spree_i18n/pull/2

This PR also needs the following PRs: #5870, #5980, #5928, #6089

This PR also needs key PR #5885 that bring product/variant models from spree.

#### What should we test?
I think paypal is the only manual test we need to do to verify the changes done in this PR.
BUT, this PR is removing spree_core all together: it's one of the biggest changes ever in OFN.

I think we should review, test and merge all the PRs mentioned above before we get to this one.

#5885 is also a big change so I think we should test it together with this PR in one single testing effort to validate OFN without Spree.

#### Release notes
Changelog Category: Removed
Remove Spree as a dependency.

### CODE REVIEW

File: spec/models/spree/product_spec.rb
Comment (Matt-Yorkley) [inline spec/models/spree/product_spec.rb]: :see_no_evil:

File: config/application.rb
Comment (luisramos0) [inline config/application.rb:28]: I'd really like to be able to have this in test.rb but when test.rb runs Spree::Core is not initialized yet, I think because config/initializers/spree.rb has not run yet. This after_initialize here works well but this not not nice:
`if Rails.env = 'test'`

File: Gemfile
Comment (Matt-Yorkley) [inline Gemfile:65]: We still use this in OFN code, right? The comment is valid.

**Review by Matt-Yorkley — APPROVED** [review]

File: Gemfile
Comment (luisramos0) [reply Gemfile:65]: as far as I understand the comment is valid in so far as it's a direct dependency of OFN.
But because we dont have spree anymore I dont think we need to mention history and most other dependencies are also direct dependencies, so I thought the comment was not needed now. It's just a good/normal dependency.

**Comment by luisramos0** [comment]
only the very flaky spec tag_rules is broken here (I have now rerun the build to see if goes green). Anyway we are good for testing :tada:

**Comment by Matt-Yorkley** [comment]
@luisramos0 maybe we should change `'acts_as_list', '= 0.2.0'` to `'acts_as_list', '~> 0.3.0'` in this branch before testing it, otherwise there will still be one or two bits that are potentially broken, no?

**Comment by luisramos0** [comment]
I think you are mixing things Matt. This is the end of bye bye spree. This is not related to the rails 4.1 upgrade where that act_as_list thing will be needed.

I think we should go ahead with this as is and afterwards, we can upgrade act_as_list for the rails upgrade :+1:

**Comment by Matt-Yorkley** [comment]
Wrong branch :see_no_evil:

**Comment by luisramos0** [comment]
ok :+1: this is either waiting for a second review or you moving it to Test Ready.

**Comment by filipefurtad0** [comment]
Hey Luis,
I've added the feedback-needed label on this one - we are experiencing deployment issues, see:

- https://openfoodnetwork.slack.com/archives/CEF14NU3V/p1603734485087600
- PR #6257

**Comment by luisramos0** [comment]
I am trying to see what's the problem in the deployment. Looks like a conflict between ofn routes and devise routes:

The error comes from this devise code:
```
  resource :session, only: [], controller: controllers[:sessions], path: "" do
          get   :new,     path: mapping.path_names[:sign_in],  as: "new"
```       
new_spree_user_session is not in ofn. In ofn we only have:
```
  devise_scope :spree_user do
    post '/login' => 'user_sessions#create', :as => :create_new_session
```
this is the full devise routes config we have on our side:
image

**Comment by Matt-Yorkley** [comment]
I think `#devise_for` automagically creates a number of routes here, and one of them is `new_spree_user_session`. Snippet from the comments in the source code:
```
    # This method is going to look inside your User model and create the
    # needed routes:
    #
    #  # Session routes for Authenticatable (default)
    #       new_user_session GET    /users/sign_in                    {controller:"devise/sessions", action:"new"}
    #           user_session POST   /users/sign_in                    {controller:"devise/sessions", action:"create"}
    #   destroy_user_session DELETE /users/sign_out                   {controller:"devise/sessions", action:"destroy"}
    #
...
```

But I don't think that's actually the problem. I think we're only seeing the error here because it's the first route in this page. The issue is that these routes are being loaded twice (this used to be silently ignored but fairly recently (Rails 4.0) changed to being a fatal error). So I think we're seeing the error on this specific route because it's the first duplicate route encountered. If we comment out this route I think we'll see the same error on the next route in the file, and so on...

The branch contains changes to the way we load the routes in `config/routes/spree`:

https://github.com/openfoodfoundation/openfoodnetwork/pull/5978/commits/21fd929e41e4f9b5e46cce93215a13486f280888
https://github.com/openfoodfoundation/openfoodnetwork/pull/5978/commits/cdf4c795c7c7ae18006eb24af9336941361b6b81

**Comment by luisramos0** [comment]
I have just spent 2 hours trying to figure this out.

If I revert this commit https://github.com/openfoodfoundation/openfoodnetwork/commit/cdf4c795c7c7ae18006eb24af9336941361b6b81 the app will not start with the same error as delayed_job above:
https://openfoodnetwork.slack.com/archives/CEF14NU3V/p1603734485087600

**Comment by luisramos0** [comment]
Spree engine and it's routes are still there. So we append to it. Although in my last version I am using draw in routes/spree.rb and append everywhere else.
My first attempt at this was to kill the spree engine and its routes and make all ofn routes but devise didnt work, we should do this at some point. We need to figure out devise routes config when we do this. Ideally we would even remove the spree_user name and call it ofn_user or even just user.

I think I have a clue now, will try in staging... I think the problem could be on better_paypal_express gem that calls routes draw instead of append... will post results here.

**Comment by luisramos0** [comment]
ok, the deployment is fixed.
The next step is to review and merge this PR that removes routes from better_spree_paypal_express: 
https://github.com/openfoodfoundation/better_spree_paypal_express/pull/10
After that I can update this bye bye spree PR to use that new version and add the routes in OFN and we can test the deployment and paypal :+1:

**Review by Matt-Yorkley — APPROVED** [review]
LGTM :ok_hand:

**Comment by luisramos0** [comment]
ok, thanks @Matt-Yorkley 
I have merged openfoodfoundation/better_spree_paypal_express#10 and I've redone the last commit in this PR to pick up that new version of better_spree_paypal_express.
The build is green and deployment is green as well.

I did a quick check and it's not looking good: all shops are closed but then I see shops with products:
https://staging.openfoodnetwork.org.uk/pepe-s-arepas-place/shop
I'll investigate this issue now and report back here :+1:

**Comment by Matt-Yorkley** [comment]
Looking at the shops page... the shops are not shown but if you click "show closed shops" it shows all the shops in open state. There's javascript console errors as well. I think it's some weird issue with one of the endpoints that fetches open and closed shops...

File: config/application.rb
Comment (Matt-Yorkley) [inline config/application.rb:28]: :joy:

Comment (luisramos0) [reply config/application.rb:28]: such a classic problem!

**Comment by filipefurtad0** [comment]
Ok Luis,

I had a fist go at the suggested test cases.

From this issue: 

**Paypal**

- verified the settings were kept, after deploy of the PR
- verified checkout is possible with this payment method (with the existing settings)
- created a "fresh" hub, products, OC and paypal payment method
- verified checkout is possible with this payment method (with the newly created hub)

All good here.



From #5885:

**Catalog management - create edit delete products and variants and all their properties**

- creating/editing variants - all good
- deleting variants: re-encountered bug #5947. Also, after deleting the last item in an order cycle, I verified the shop to be open anyway:

image

I'm not sure this is a bug - observed in other staging servers, so not introduced by this PR (staging-es below):

image

- add these to OCs - OK
- verify we can buy these products and variants and that they display correctly in the shopfront - OK 
- verify reports and make sure products and variants are displaying correctly - OK

This is the first stage of testing, found nothing unusual.

> there quite a few more things we could check...

Yes, I agree. I will pick up this again later on, and consider different test cases, and follow a similar approach as with the other "OFN heart surgery" PRs, i.e., try to reproduce as many related bugs as possible, to assure the app is working as before. More on this later. I'll remove the staging label for now.

**Comment by Matt-Yorkley** [comment]
@luisramos0 I think the target branch is behind master here, it's not great great for testing...

**Comment by luisramos0** [comment]
I thought the deploy process included a merge of master into the PR?
I rebased it anyway now.

**Comment by filipefurtad0** [comment]
Hey @luisramos0 ,

I had a second round on testing here, looking for issues and test cases around "Prices". Other than the Cancel / Resume button (via DM) and on the notes below, I found the app to be working as expected.
https://docs.google.com/document/d/1HJ02BHX36Vjj6suKFAEqOryT7I5hC1bN9On4ApXnnlg/edit#

I'll continue testing this tomorrow, and address "products/variants", for example as well.

**Comment by Matt-Yorkley** [comment]
This is a tense one! :smile:

Suspense!

**Comment by luisramos0** [comment]
@Matt-Yorkley mentioned some problems with merging master with this PR so I just rebased this branch again with latest master.

**Comment by luisramos0** [comment]
hey @filipefurtad0 the cancel button issue was a simple missing translation so I added a commit to this PR to fix it.
You can restage and have a look :+1:

**Comment by filipefurtad0** [comment]
Awesome, thanks @luisramos0 !
In the meantime, I found something quite similar on the OC date pickers. The missing translation and the strangely looking arrows:

image

I've been updating the test file indicated previously :+1:  but maybe I should stop soon - there are plenty of product/variant related issues, I don't think I'll reproduce them all. I found other bugs but not related to the byebyeSpree, so I think this looks pretty good to go.

So, maybe I'll wait for this latest translation fix and re-stage after rebuild, for a final checkup.

**Comment by luisramos0** [comment]
ok Filipe, I have fixed that in a new commit. Missing translations.

It's likely that we will find a few more translations missing, we are leaving behind a big spree translations file:
https://github.com/openfoodfoundation/spree/blob/master/core/config/locales/en.yml
Some of these are not used in OFN, some are and are already copied to our en.yml file but there maybe some of them missing like the cancel button above or the order cycle next button.

**Comment by filipefurtad0** [comment]
Ok Luis, I'd say this is good to go. The app behaves as before!! :tada: 
Or at least I haven't found any reason to believe otherwise :+1: 

- List of issues reproduced:

Bug if variant price is updated while it's already in a cart #1818 
Rounding issues / decimals in orders#6261
Euro currency error. When I type a price of 5 it turns into 50 #3057
In backoffice orders, the set enterprise fees are not included in the price per item #5567
Bugs in the Products and Inventory reports #2529 (some)
S3-bug -Invalid input in the "Bulk Edit Products" should not update any record #2774
Product duplication has errors or not working at all #660
Summary calculator box not populating in BOM for 'item' products #664
Weight Calculator: If an 'item' product has previously been sold as 'g/kg', it will incorrectly attract the fee.#3308
Order cycle selector appears to but DOESN'T select OC and no products load #5100
Flat percent calculator in shipping/payment method only applies to products total (excl. ent. fee) #2003

- List of opened issues (not introduced by this PR):
Shipping fees: Can't change from weight calculator (error 500/snail) #6300
Product edit: Cannot change Variant unit scale of an existing product #6304

- Found another missing translation with this PR:
image

Should I open an issue as usual Luis?
  
- A rollback scenario was tested upon release testing, as discussed [here](https://openfoodnetwork.slack.com/archives/C02TZ6X00/p1603980838100700) and in the following threads. I checked the app as well, for previous settings on "my" shops, found nothing unusual.

Good to go!!
Godspeed "bye bye Spree"!

**Comment by luisramos0** [comment]
I cant believe I just merged this!
Thanks a lot Filipe!

**Comment by luisramos0** [comment]
@filipefurtad0 missing translation fix is here: https://github.com/openfoodfoundation/openfoodnetwork/pull/6308

**Pull request merged**
